pkg/ccl/changefeedccl/changefeed_test.go

```diff
@@ -3941,3 +3945,143 @@ func TestChangefeedCheckpointSchemaChange(t *testing.T) {
+				// we want to resolve those sub-spans instead of the entire table span.
+				// However, we have to emit something -- otherwise the entire changefeed
+				// machine would not work.
+				r.Span.EndKey = tableSpan.Key.Next()
+				return true
+			}
+			if haveGaps {
+				return rnd.Intn(10) < 7
+			}
+			haveGaps = true
+			return false
+		}
+
+		// Checkpoint progress frequently.
+		changefeedbase.FrontierCheckpointFrequency.Override(
+			context.Background(), &f.Server().ClusterSettings().SV, 10*time.Millisecond)
+
+		registry := f.Server().JobRegistry().(*jobs.Registry)
+		foo := feed(t, f, `CREATE CHANGEFEED FOR foo WITH resolved='100ms'`)
+		// Some test feeds (kafka) are not buffered, so we have to consume messages.
+		g := ctxgroup.WithContext(context.Background())
+		g.Go(func() error {
+			for {
+				_, err := foo.Next()
+				if err != nil {
+					return err
+				}
+			}
+		})
+		defer func() {
+			closeFeed(t, foo)
+			_ = g.Wait()
+		}()
+
+		jobFeed := foo.(cdctest.EnterpriseTestFeed)
+		loadProgress := func() jobspb.Progress {
+			jobID := jobFeed.JobID()
+			job, err := registry.LoadJob(context.Background(), jobID)
+			require.NoError(t, err)
+			return job.Progress()
+		}
+
+		// Wait for non-nil checkpoint.
+		testutils.SucceedsSoon(t, func() error {
+			progress := loadProgress()
+			if p := progress.GetChangefeed(); p != nil && p.Checkpoint != nil && len(p.Checkpoint.Spans) > 0 {
+				return nil
+			}
+			return errors.New("waiting for checkpoint")
+		})
+
+		// Pause the job and read and verify the latest checkpoint information.
+		require.NoError(t, jobFeed.Pause())
+		progress := loadProgress()
+		require.NotNil(t, progress.GetChangefeed())
+		h := progress.GetHighWater()
+		noHighWater := (h == nil || h.IsEmpty())
+		require.True(t, noHighWater)
+		checkpoint := progress.GetChangefeed().Checkpoint
+		require.Less(t, 0, len(checkpoint.Spans))
+		checkpointFrontier, err := span.MakeFrontier(checkpoint.Spans...)
+
+		require.NoError(t, err)
+
+		// Collect spans we attempt to resolve after when we resume.
+		var resolved []roachpb.Span
+		knobs.ShouldEmitResolved = func(r *jobspb.ResolvedSpan) bool {
+			if !r.Span.Equal(tableSpan) {
+				resolved = append(resolved, r.Span)
+			}
+			return true
+		}
+
+		// Resume job.
+		require.NoError(t, jobFeed.Resume())
+
+		// Wait for high water mark to be non-zero
+		testutils.SucceedsSoon(t, func() error {
+			progress := loadProgress()
+			if p := progress.GetHighWater(); p != nil && !p.IsEmpty() {
+				return nil
+			}
+			return errors.New("waiting for highwater")
+		})
+
+		// At this point, highwater mark should be set, and previous checkpoint should be gone.
+		progress = loadProgress()
+		require.NotNil(t, progress.GetChangefeed())
+		require.Equal(t, 0, len(progress.GetChangefeed().Checkpoint.Spans))
+
+		// Verify that none of the resolved spans after resume were checkpointed.
+		for _, sp := range resolved {
```

Comment:
This feels like another place where the `span.Frontier` is overkill and a `SpanGroup` will be cleaner.